File: front-end/src/types/storeMaterial.ts
/**
 * 가맹점 재료 도메인 공통 타입 정의
 *
 * - 백엔드 Material / StoreMaterial 엔티티와 1:1은 아니지만,
 *   주요 Enum/DTO 구조/명세를 그대로 맞춘다.
 * - 수량/금액의 단위 기준에 유의:
 *   - 수량: 기본적으로 "소진 단위"
 *   - 금액: 기본적으로 "입고 단위(매입 단가)"
 */

/** 가맹점 재료 상태 */
export type MaterialStatus = 'USE' | 'STOP';

/** StoreMaterialStatus 는 MaterialStatus 의 별칭(동일 의미) */
export type StoreMaterialStatus = MaterialStatus;

/** 보관 온도 */
export type MaterialTemperature = 'TEMPERATURE' | 'REFRIGERATE' | 'FREEZE';

/** 재료 카테고리 (본사 MaterialCategory Enum과 동일) */
export type MaterialCategory =
  | 'BASE'
  | 'TOPPING'
  | 'SIDE'
  | 'SAUCE'
  | 'BEVERAGE'
  | 'PACKAGE'
  | 'ETC';

/**
 * 가맹점 재료 등록 요청 DTO
 *
 * - 엔드포인트: POST /API/store/material
 * - 백엔드 StoreMaterialCreateDTO 와 필드명을 최대한 맞춘다.
 * - 단위/환산:
 *   - baseUnit: 소진 단위
 *   - salesUnit: 입고 단위
 *   - conversionRate: 입고→소진 환산 (예: 1박스=100ea → 100)
 */
export interface StoreMaterialCreateRequest {
  /** 본사 재료 매핑 여부 (true=본사 재료, false=가맹점 자체 재료) */
  hqMaterial?: boolean | null;

  /** 본사 재료 PK (hq material id). hqMaterial=true 일 때 의미 있음 */
  materialId?: number | null;

  /** 가맹점 표시명 (재료명) */
  name: string;

  /** 카테고리 (MaterialCategory) */
  category?: MaterialCategory | null;

  /** 소진 단위 (예: 개, g, 샷) */
  baseUnit?: string | null;

  /** 입고 단위 (예: 박스, 봉, kg) */
  salesUnit?: string | null;

  /** 변환비율(입고→소진). 예: 1박스=100ea → 100 */
  conversionRate?: number | null;

  /** 공급업체명 (대표/최근 공급처) */
  supplier?: string | null;

  /** 보관온도 */
  temperature?: MaterialTemperature | null;

  /** 적정 재고(소진 단위) */
  optimalQuantity?: number | null;

  /** 최근 매입 단가(입고 단위 기준) */
  purchasePrice?: number | null;
}

/**
 * 가맹점 재료 목록/상세 응답 DTO
 *
 * - 엔드포인트: GET /API/store/material/list
 * - 다른 화면에서도 재사용 가능하도록 넉넉한 필드 포함
 */
export interface StoreMaterialResponse {
  /** 가맹점 재료 PK (store_material_id) */
  id: number;

  /** 가맹점 ID (store_id_fk) */
  storeId: number;

  /** 가맹점 재료 코드 */
  code: string;

  /** 가맹점 재료명(표시명) */
  name: string;

  /** 카테고리 */
  category: MaterialCategory | null;

  /** 소진 단위 */
  baseUnit: string | null;

  /** 입고 단위 */
  salesUnit: string | null;

  /** 재료 상태 (USE/STOP) */
  status: MaterialStatus;

  /** 적정 재고(소진 단위) */
  optimalQuantity: number | null;

  /** 최근 매입 단가(입고 단위 기준) */
  purchasePrice: number | null;

  /** 공급업체명 */
  supplier: string | null;

  /** 보관온도 */
  temperature: MaterialTemperature | null;

  /** 본사 재료 여부 (true=본사 재료, false=가맹점 자체 재료) */
  hqMaterial: boolean;
}

/**
 * 가맹점 재료 적정재고(최소 재고) 업데이트 DTO
 *
 * - 목적: 상세 팝업에서 최소 재고(적정재고) 저장
 * - 비고: null 허용(미설정), 소진 단위 기준
 */
export interface StoreMaterialUpdateOptimalDTO {
  /** 가맹점 재료 PK (store_material_id) */
  storeMaterialId: number;

  /** 적정 재고 수량(소진 단위). null 이면 미설정 */
  optimalQuantity: number | null;
}

/**
 * 가맹점 재료 상태(사용/중지) 업데이트 DTO
 *
 * - 목적: 본사 재료 사용 토글 on/off
 * - 비고: 본사 재료가 아닌 경우 토글을 숨기거나 요청 자체를 막는 편이 안전
 */
export interface StoreMaterialUpdateStatusDTO {
  /** 가맹점 재료 PK (store_material_id) */
  storeMaterialId: number;

  /** 'USE' | 'STOP' */
  status: StoreMaterialStatus;
}

File: src/main/java/com/boot/ict05_final_user/domain/inventory/entity/StoreMaterial.java
package com.boot.ict05_final_user.domain.inventory.entity;

import com.boot.ict05_final_user.domain.store.entity.Store;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Comment;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * 가맹점 재료(StoreMaterial) 엔티티
 *
 * <p>
 * - 재고 수량/유통기한/입출고 이력은 {@link StoreInventory} 및 이후
 *   StoreInventoryBatch/Log 에서만 관리한다.<br>
 * - 이 엔티티는 “가맹점 기준 재료 마스터 + 적정 재고 + 최근 매입 정보”까지만 가진다.
 * </p>
 *
 * <ul>
 *   <li>본사 공급 재료(hqMaterial = true) → material FK 존재</li>
 *   <li>가맹점 자체 등록 재료(hqMaterial = false) → material FK = NULL</li>
 * </ul>
 */
@Entity
@Table(
        name = "store_material",
        uniqueConstraints = @UniqueConstraint(
                name = "uq_store_material_code",
                columnNames = {"store_id_fk", "store_material_code"}
        )
)
@Getter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class StoreMaterial {

    /** 가맹점 재료 고유 ID */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "store_material_id", columnDefinition = "BIGINT UNSIGNED")
    @Comment("가맹점 재료 시퀀스")
    private Long id;

    /** 가맹점 (FK: store.store_id) */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(
            name = "store_id_fk",
            nullable = false,
            foreignKey = @ForeignKey(name = "fk_sm_store"),
            columnDefinition = "BIGINT UNSIGNED"
    )
    @Comment("매장 시퀀스 (FK)")
    private Store store;

    /** 본사 재료 (FK: material.material_id) – 본사 재료 사용 시에만 설정 */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(
            name = "material_id_fk",
            foreignKey = @ForeignKey(name = "fk_sm_material"),
            columnDefinition = "BIGINT UNSIGNED"
    )
    @Comment("본사 재료 (FK)")
    private Material material;

    /** 가맹점 재료 코드 (점포별 고유) */
    @Column(name = "store_material_code", length = 30, nullable = false,
            columnDefinition = "VARCHAR(30)")
    @Comment("가맹점 재료 코드(점포별 고유)")
    private String code;

    /** 가맹점 재료명 (표시명) */
    @Column(name = "store_material_name", length = 100, nullable = false,
            columnDefinition = "VARCHAR(100)")
    @Comment("가맹점 재료명")
    private String name;

    /** 카테고리 (본사의 MaterialCategory Enum 값을 문자열로 저장) */
    @Column(name = "store_material_category", length = 50,
            columnDefinition = "VARCHAR(50)")
    @Comment("가맹점 재료 카테고리")
    private String category;

    /** 소진 단위 (가맹점 기준 – 예: 개, 샷, g) */
    @Column(name = "store_material_base_unit", length = 20,
            columnDefinition = "VARCHAR(20)")
    @Comment("소진 단위(가맹점 기준)")
    private String baseUnit;

    /** 입고 단위 (본사 판매단위/발주단위 – 예: 박스, 통, kg) */
    @Column(name = "store_material_sales_unit", length = 20,
            columnDefinition = "VARCHAR(20)")
    @Comment("입고 단위(본사 기준 단위)")
    private String salesUnit;

    /** 변환비율(입고단위 → 소진단위, 예: 1박스=1000g) */
    @Column(name = "material_conversion_rate", nullable = false,
            columnDefinition = "INT DEFAULT 1")
    @Comment("변환비율(입고단위 → 소진단위)")
    private Integer conversionRate;

    /** 공급업체명 (최근/대표 공급처) */
    @Column(name = "store_material_supplier", length = 100,
            columnDefinition = "VARCHAR(100)")
    @Comment("가맹점 재료 공급업체명(대표/최근)")
    private String supplier;

    /** 보관온도 */
    @Enumerated(EnumType.STRING)
    @Column(name = "store_material_temperature",
            columnDefinition = "ENUM('TEMPERATURE','REFRIGERATE','FREEZE')")
    @Comment("보관온도")
    private MaterialTemperature temperature;

    /** 재료 상태 (USE=사용, STOP=미사용) */
    @Setter
    @Enumerated(EnumType.STRING)
    @Column(name = "store_material_status", nullable = false,
            columnDefinition = "ENUM('USE','STOP') DEFAULT 'USE'")
    @Comment("재료 상태")
    private MaterialStatus status;

    /** 적정 수량 (가맹점 기준 – 보통 소진단위 기준으로 운용) */
    @Setter
    @Column(name = "store_material_optimal_quantity", precision = 15, scale = 3,
            columnDefinition = "DECIMAL(15,3)")
    @Comment("적정 수량(가맹점 기준)")
    private BigDecimal optimalQuantity;

    /** 최근 매입 단가 (입고단위 기준 금액) */
    @Column(name = "store_material_purchase_price",
            columnDefinition = "DECIMAL(15,2)")
    @Comment("최근 매입 단가(입고단위 기준)")
    private BigDecimal purchasePrice;

    /** 본사 재료 여부 (1=본사, 0=가맹점 자체 등록) */
    @Column(name = "store_material_is_hq_material", nullable = false,
            columnDefinition = "TINYINT(1) DEFAULT 0")
    @Comment("본사 재료 여부")
    private boolean isHqMaterial;

    /** 등록일 */
    @CreationTimestamp
    @Column(name = "store_material_reg_date", nullable = false,
            columnDefinition = "DATETIME DEFAULT CURRENT_TIMESTAMP")
    @Comment("등록일")
    private LocalDateTime regDate;

    /** 수정일 */
    @UpdateTimestamp
    @Column(name = "store_material_modify_date",
            columnDefinition = "DATETIME DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP")
    @Comment("수정일")
    private LocalDateTime modifyDate;
}
